File: totoggle_go/internal/cache/cache_test.go
package cache

import (
	"errors"
	"sync"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/manorfm/toToggles/totoggle_go/internal/toggle"
)

func mustPath(t *testing.T, raw string) toggle.Path {
	t.Helper()
	p, err := toggle.NewPath(raw)
	require.NoError(t, err)
	return p
}

func TestCache_Update_StoresApplicationAndRecordsSuccess(t *testing.T) {
	c := New()
	t1 := toggle.Toggle{ID: "1", Path: mustPath(t, "t1"), Enabled: true}

	before := time.Now()
	c.Update(toggle.NewApplication([]toggle.Toggle{t1}))
	after := time.Now()

	stats := c.Stats()
	assert.Equal(t, 1, stats.ToggleCount)
	assert.False(t, stats.LastSuccessAt.Before(before))
	assert.False(t, stats.LastSuccessAt.After(after))
	assert.Zero(t, stats.ConsecutiveFailures)
}

func TestCache_Get_ReturnsTargetAndAncestors(t *testing.T) {
	c := New()
	t1 := toggle.Toggle{ID: "1", Path: mustPath(t, "t1"), Enabled: true}
	t1t2 := toggle.Toggle{ID: "2", Path: mustPath(t, "t1.t2"), Enabled: false}
	c.Update(toggle.NewApplication([]toggle.Toggle{t1, t1t2}))

	target, ancestors, ok := c.Get(mustPath(t, "t1.t2"))
	require.True(t, ok)
	assert.Equal(t, t1t2, target)
	require.Len(t, ancestors, 1)
	assert.Equal(t, t1, ancestors[0])
}

func TestCache_Get_NotFound(t *testing.T) {
	c := New()
	c.Update(toggle.NewApplication(nil))

	_, _, ok := c.Get(mustPath(t, "missing"))
	assert.False(t, ok)
}

func TestCache_Get_BeforeAnyUpdate(t *testing.T) {
	c := New()

	_, _, ok := c.Get(mustPath(t, "t1"))
	assert.False(t, ok)
}

// A failed refresh must never blank out previously-fetched data — stale-but-present beats empty.
func TestCache_RecordFailure_KeepsPriorDataAndTracksFailure(t *testing.T) {
	c := New()
	t1 := toggle.Toggle{ID: "1", Path: mustPath(t, "t1"), Enabled: true}
	c.Update(toggle.NewApplication([]toggle.Toggle{t1}))

	failErr := errors.New("server unreachable")
	c.RecordFailure(failErr)
	c.RecordFailure(failErr)

	target, _, ok := c.Get(mustPath(t, "t1"))
	require.True(t, ok)
	assert.Equal(t, t1, target)

	stats := c.Stats()
	assert.Equal(t, 2, stats.ConsecutiveFailures)
	assert.Equal(t, failErr, stats.LastError)
	assert.False(t, stats.LastErrorAt.IsZero())
}

func TestCache_Update_ResetsConsecutiveFailures(t *testing.T) {
	c := New()
	c.RecordFailure(errors.New("boom"))
	c.RecordFailure(errors.New("boom"))

	c.Update(toggle.NewApplication(nil))

	assert.Zero(t, c.Stats().ConsecutiveFailures)
}

func TestCache_Clear_RemovesDataAndResetsStats(t *testing.T) {
	c := New()
	t1 := toggle.Toggle{ID: "1", Path: mustPath(t, "t1"), Enabled: true}
	c.Update(toggle.NewApplication([]toggle.Toggle{t1}))
	c.RecordFailure(errors.New("boom"))

	c.Clear()

	_, _, ok := c.Get(mustPath(t, "t1"))
	assert.False(t, ok)
	assert.Equal(t, Stats{}, c.Stats())
}

func TestCache_ConcurrentAccess(t *testing.T) {
	c := New()
	var wg sync.WaitGroup
	for i := 0; i < 50; i++ {
		wg.Add(3)
		go func() {
			defer wg.Done()
			c.Update(toggle.NewApplication([]toggle.Toggle{{ID: "1", Path: mustPath(t, "t1")}}))
		}()
		go func() {
			defer wg.Done()
			c.Get(mustPath(t, "t1"))
		}()
		go func() {
			defer wg.Done()
			c.RecordFailure(errors.New("boom"))
		}()
	}
	wg.Wait()
}

File: totoggle_node/src/internal/cache/cache.test.ts
import { describe, expect, it } from "vitest";
import { Path } from "../toggle/path.js";
import { Application } from "../toggle/application.js";
import type { Toggle } from "../toggle/toggle.js";
import { Cache } from "./cache.js";

function toggle(path: string, overrides: Partial<Toggle> = {}): Toggle {
  return {
    id: path,
    path: Path.parse(path),
    value: path,
    enabled: true,
    level: 0,
    parentId: null,
    appId: "app-1",
    hasActivationRule: false,
    activationRule: null,
    ...overrides,
  };
}

describe("Cache", () => {
  it("uses the injected clock for deterministic refresh timestamps", () => {
    const now = new Date("2026-01-02T03:04:05.000Z");
    const cache = new Cache(() => now);

    cache.update(new Application([]));

    expect(cache.stats().lastSuccessAt).toEqual(now);
  });

  it("stores an Application and records success stats on update", () => {
    const cache = new Cache();
    const before = Date.now();

    cache.update(new Application([toggle("t1")]));

    const stats = cache.stats();
    expect(stats.toggleCount).toBe(1);
    expect(stats.lastSuccessAt).not.toBeNull();
    expect(stats.lastSuccessAt!.getTime()).toBeGreaterThanOrEqual(before);
    expect(stats.consecutiveFailures).toBe(0);
  });

  it("keeps the ETag and revision with the current snapshot", () => {
    const cache = new Cache();

    cache.update(new Application([toggle("t1")]), { etag: '"catalog-v1"', revision: "1" });

    expect(cache.catalog()).toEqual({ etag: '"catalog-v1"', revision: "1" });
  });

  it("treats a 304 as success without replacing the cached snapshot", () => {
    const cache = new Cache();
    const app = new Application([toggle("t1")]);
    cache.update(app, { etag: '"catalog-v1"', revision: "1" });
    cache.recordFailure(new Error("temporary failure"));

    cache.markNotModified({ etag: '"catalog-v1"' });

    expect(cache.get(Path.parse("t1"))?.target.id).toBe("t1");
    expect(cache.catalog()).toEqual({ etag: '"catalog-v1"', revision: "1" });
    expect(cache.stats().consecutiveFailures).toBe(0);
  });

  it("keeps the prior ETag when a 304 omits an ETag header", () => {
    const cache = new Cache();
    cache.update(new Application([toggle("t1")]), { etag: '"catalog-v1"' });

    cache.markNotModified({ etag: undefined });

    expect(cache.catalog()).toEqual({ etag: '"catalog-v1"' });
  });

  it("get returns the target and its ancestors", () => {
    const cache = new Cache();
    const t1 = toggle("t1");
    const t1t2 = toggle("t1.t2", { enabled: false });
    cache.update(new Application([t1, t1t2]));

    const result = cache.get(Path.parse("t1.t2"));
    expect(result).not.toBeNull();
    expect(result!.target).toEqual(t1t2);
    expect(result!.ancestors).toEqual([t1]);
  });

  it("get returns null for a path never fetched", () => {
    const cache = new Cache();
    cache.update(new Application([]));

    expect(cache.get(Path.parse("missing"))).toBeNull();
  });

  it("get returns null before any update", () => {
    const cache = new Cache();
    expect(cache.get(Path.parse("t1"))).toBeNull();
  });

  // A failed refresh must never blank out previously-fetched data — stale-but-present beats
  // empty.
  it("recordFailure keeps prior data and tracks the failure", () => {
    const cache = new Cache();
    cache.update(new Application([toggle("t1")]));

    const failure = new Error("server unreachable");
    cache.recordFailure(failure);
    cache.recordFailure(failure);

    expect(cache.get(Path.parse("t1"))!.target.id).toBe("t1");
    const stats = cache.stats();
    expect(stats.consecutiveFailures).toBe(2);
    expect(stats.lastError).toBe(failure);
    expect(stats.lastErrorAt).not.toBeNull();
  });

  it("update resets the consecutive failure count", () => {
    const cache = new Cache();
    cache.recordFailure(new Error("boom"));
    cache.recordFailure(new Error("boom"));

    cache.update(new Application([]));

    expect(cache.stats().consecutiveFailures).toBe(0);
  });

  it("clear removes data and resets stats", () => {
    const cache = new Cache();
    cache.update(new Application([toggle("t1")]));
    cache.recordFailure(new Error("boom"));

    cache.clear();

    expect(cache.get(Path.parse("t1"))).toBeNull();
    const stats = cache.stats();
    expect(stats.toggleCount).toBe(0);
    expect(stats.lastSuccessAt).toBeNull();
    expect(stats.consecutiveFailures).toBe(0);
  });

  // Node is single-threaded, but fetches are async and can interleave — a failure recorded from
  // an in-flight refresh must never race ahead of a later, already-applied success.
  it("interleaved update/recordFailure calls never corrupt the toggle data", async () => {
    const cache = new Cache();
    const ops = Array.from({ length: 50 }, (_, i) =>
      i % 2 === 0
        ? Promise.resolve().then(() => cache.update(new Application([toggle("t1")])))
        : Promise.resolve().then(() => cache.recordFailure(new Error("boom"))),
    );
    await Promise.all(ops);

    const result = cache.get(Path.parse("t1"));
    expect(result === null || result.target.id === "t1").toBe(true);
  });
});
